src/Dom/Document/Filter/ProtectEsiTags.php

```diff
@@ -0,0 +1,61 @@
+<?php
+
+namespace AmpProject\Dom\Document\Filter;
+
+use AmpProject\Dom\Document\AfterSaveFilter;
+use AmpProject\Dom\Document\BeforeLoadFilter;
+
+/**
+ * Protect the esi tags from being broken.
+ *
+ * @package ampproject/amp-toolbox
+ */
+final class ProtectEsiTags implements BeforeLoadFilter, AfterSaveFilter
+{
+
+    /**
+     * List of self closing ESI tags.
+     *
+     * @link https://www.w3.org/TR/esi-lang/
+     *
+     * @var string[]
+     */
+    const SELF_CLOSING_TAGS = [
+        'esi:include',
+        'esi:comment',
+    ];
+
+    /**
+     * Preprocess the HTML to be loaded into the Dom\Document.
+     *
+     * @param string $html String of HTML markup to be preprocessed.
+     * @return string Preprocessed string of HTML markup.
+     */
+    public function beforeLoad($html)
+    {
+        $selfClosingregex = '#<(' . implode('|', self::SELF_CLOSING_TAGS) . ')([^>]*?)(?>\s*(?<!\\\\)/)?>(?!</\1>)#';
+
+        $html = preg_replace($selfClosingregex, '<$1$2></$1>', $html);
```

Comment:
Given that the subsequent replaces don't act on full open&close tags, but only on partials, I don't understand what the problem was with the self-closing tags in the first place.

Were they being stripped to enable one of the subsequent transformations?

Reply:
@schlessera Should I add the tags in `\AmpProject\Tag::SELF_CLOSING_TAGS` list? I just wanted to keep them separate from the original ones.

Reply:
I just tested, and the issue with the self-closing tags is that `DOMDocument` adds things like `</esi:include>` as additional tags for these.